jf_agent/main.py

```diff
@@ -528,28 +261,62 @@ def obtain_jellyfish_endpoint_info(config, creds):
     jira_info = agent_config.get('jira_info')
     git_instance_info = agent_config.get('git_instance_info')
 
-    if config.git_url and len(git_instance_info) != 1:
+    # if no git info has returned from the endpoint, then an instance may not have been provisioned
+    if len(config.git_configs) > 0 and not len(git_instance_info.values()):
         print(
-            f'ERROR: Invalid Git instance info returned from the agent config endpoint -- please contact Jellyfish'
+            f'ERROR: A git instance is configured, but no git instance '
+            f'info returned from the jellyfish api -- please contact Jellyfish'
         )
         raise BadConfigException()
 
+    # if there are multiple git configurations
+    if len(config.git_configs) > 1:
+        for git_config in config.git_configs:
+            # assert that each instance in the config is mappable to the instances
+            # returned by the endpoint (we'll need this data later)
+            slug = git_config.git_instance_slug
+            if not git_instance_info.get(slug):
+                print(
+                    f'ERROR: The jellyfish api did not return an instance with the git_instance_slug `{slug}` -- '
+                    f'please check your configuration or contact Jellyfish'
+                )
+                raise BadConfigException()
+
     return JellyfishEndpointInfo(jira_info, git_instance_info)
 
 
 @diagnostics.capture_timing()
 @agent_logging.log_entry_exit(logger)
-def download_data(
-    config, endpoint_jira_info, endpoint_git_instance_info, jira_connection, git_connection
-):
+def download_data(config, creds, endpoint_jira_info, endpoint_git_instances_info):
     download_data_status = []
 
-    if jira_connection:
+    if config.jira_url:
+        jira_connection = get_basic_jira_connection(config, creds)
+        if config.run_mode_is_print_all_jira_fields:
+            print_all_jira_fields(config, jira_connection)
         download_data_status.append(load_and_dump_jira(config, endpoint_jira_info, jira_connection))
 
-    if git_connection:
+    is_multi_git_config = len(config.git_configs) > 1
+    for git_config in config.git_configs:
+        git_connection = get_git_client(
+            git_config, creds, skip_ssl_verification=config.skip_ssl_verification
+        )
+
+        if is_multi_git_config:
+            instance_slug = git_config.git_instance_slug
+            instance_info = endpoint_git_instances_info.get(instance_slug)
+        else:
+            # support legacy single-git support, of which assumes only one available git instance
+            instance_info = list(endpoint_git_instances_info.values())[0]
```

Comment:
This relies on a change first having been deployed on the Jellyfish side, right?

Reply:
Nope. The shape returning from the api is the actually same, it just has more items. Grabbing the first (and only item) is what we are doing currently.